### ISSUE TITLE
Dotnet core tests freeze on VSTS build with IClassFixture implemented

### ISSUE THREAD

**Issue by Larswa** [xunit/xunit#1065]
On a small asp.net core project I am using Xunit as the test framework (default settings). I am running a continous integration build from VSTS with a hosted agent and a local build agent.

Testing has run fine for months on the CI builds, but after implementing IClassFixture the tests now hang on both build platforms (hosted and local build agents).

"dotnet test" finds the test project and fires off the correct command, but just hangs there for the next hour untill the build times out.  
`[command]C:\Program Files (x86)\dotnet\dotnet.exe test C:/Agent/_work/3/s/Test/FactusAviation.UnitTests/project.json --configuration release -xml C:\Agent\_work\3\a\TestResults.xml -trait Category=Booking`

I have tried both with and without adding traits to the test and the "dotnet test" command. No difference.
I am not getting any errors to go by ... not even on the local build server. It just hangs after firing off the dotnet test command.

The same command works fine locally om both my dev machines.
For testing purposes, I created another dotnet core MSTest project, and that runs on the CI build.

After removing the IClassFixture setup, tests run again as before, on my VSTS builds.

Any ideas?

From my project.json for the dotnet core test project: (Cherrypicked lines ... not the complete project.json)

```
"dotnet-test-xunit": "2.2.0-preview2-build1029",
"xunit": "2.2.0-beta4-build3444"

"frameworks": {
    "netcoreapp1.1": {
      "dependencies": {
        "Microsoft.NETCore.App": {
          "type": "platform",
          "version": "1.1.0"
        }
      },
      "imports": [
        "dnxcore50",
        "portable-net45+win8"
      ]
    }
},
"testRunner": "xunit",
```

**Comment by bradwilson** [comment]
Can you provide a repro project?

**Comment by Larswa** [comment]
Hi Brad, yeah. I put it online here:  https://github.com/Larswa/XunitVSTSIssue

The testproject that is failing is Test/FactusAviation.UnitTests. Right now there are only 6 tests, and one skipped test.

Should be recreatable on a Visual Studio Team System hosted build agent. I am running it from a Vnext build system with either the commandline  "Dotnet test Test/factusaviation.Unittests" og from Microsofts own dotnet build/test activity.

And on my Azure hosted build server (a Win 2012R2 server) you can have an admin account if you want. I'll probably scratch it anyway. 
There are no errors reported in the build agent logs  (the _diag folder). So I am flying a little blind.
Also ... Tried procmon but it didnt show anything suspicious.

**Comment by soloitguy** [comment]
Any resolution to this?

**Comment by DalSoft** [comment]
We had this problem on VSTS only, all four devs machines were running our tests fine.

This issue happens with `IClassFixture` and `ICollectionFixture`.

One of our team found a workaround add xunit.runner.json with the following settings
```js
{
  "parallelizeTestCollections": false,
  "maxParallelThreads": -1
}
```
Again we only had to do this on VSTS so we only added this to the CI build as it will run your tests slower.

**Comment by kamnel** [comment]
We had same issue on development machines (VS2017). Our tests were using `IClassFixture` too. @DalSoft workaround finally helped to get it working.

**Comment by dberryalliance** [comment]
I ran into the same issue with an IClassFixture, but only some _specific_ tests would cause the entire VSTS test task to lockup, and I'd have to use the `Skip` directive to prevent the tests from running.  My project is a .Net 4.5.2 Visual Studio project (not dotnet core, or VSCode).  

There didn't seem to be any rhyme, reason, or shared trait between those tests that would work in parallel and those that wouldn't.  In some cases, the tests that wouldn't work were practically identical to other tests that would.  It wasn't random tests--I could specifically identify those that wouldn't run.  However, with @DalSoft's fix, I was able to make them all run.

**Comment by bradwilson** [comment]
Setting `maxParallelThreads` to `-1` is (very) strongly indicating that you have a deadlock around `Task`s.

**Comment by bradwilson** [comment]
Don't ever, ever, ever, ever, ever call `.Wait()`, `.Result`, or .GetAwaiter().GetResult() on an incomplete `Task`. Ever. You will deadlock.

**Comment by DalSoft** [comment]
@bradwilson I can confirm that I was definitely not doing this. It was the hosted VSTS test runner causing my issue. I've since upgraded the packages and now running on the 2017 VSTS agent, and it's running perfectly every time with the same code.

If people are getting this issue take @bradwilson advice look for a deadlock, then look to see if there is a bug in the test runner your using (I've had problems with Resharper's test runner in the past too), and finally please only use my workaround as a stop gap until you can find the root cause.

**Comment by dberryalliance** [comment]
@bradwilson I am using `.Wait()` and `.Result` heavily in my tests.  Both for `Fact`s that would lock up and those that wouldn't.  From my reading, though, xUnit supports async methods for tests, so perhaps all I need to do is convert my methods over to that and try again.  I'll report back here if that helped me.  Nonetheless, I don't mean to derail from the originally reported issue with improperly structured tests.

**Comment by dberryalliance** [comment]
@bradwilson Thank you for the advice.  Moving my tests into `async` was a _little_ difficult, but not impossible.  Ended up coming up with a good pattern for a simulated async IEnumerable with C# v7 (though 8 should have it natively--based on early projections).  Nonetheless, my tests worked swimmingly.

Cheers!

**Comment by bradwilson** [comment]
Closing, because I believe everybody who expressed issues has now worked them out.

**Comment by adamkjonesmesh** [comment]
Regarding the integration tests not running on VSTS, we are experiencing the exact same issue.  All of our tests are async.  There are _no_ locks occurring, and the tests run fine on multiple local boxes.  They lock up and _never_ run on VSTS.  They were running on VSTS prior to Jan. 12, 2018.  After that (which included a MSFT update to VSTS), they stopped working.  We have not been able to get them working again.  So...I know this is closed, but this still seems to be an issue.

**Comment by Keen83** [comment]
Same problem

**Comment by dmitry-pavlov** [comment]
What @DalSoft [proposed](https://github.com/xunit/xunit/issues/1065#issuecomment-283621608) or this (actually the same):
```
{
  "parallelizeTestCollections": true,
  "maxParallelThreads": -1
}
```

**Comment by alokbaburdc** [comment]
This issue is not resolved yet. My Fixture class is tiny as possible and still it uses up 100% CPU.
A process serviceHub.Host.CLR.x86(32 bit) is using up all the cpus.

**Comment by bradwilson** [comment]
@alokbaburdc Open a new issue with a repro, please.

**Comment by alokbaburdc** [comment]
@bradwilson please check #1982. Thank you.
